### ISSUE TITLE
"Error: Assembler error" with -O3 flag on macOS  Sonoma 14.7.1

### ISSUE THREAD

**Issue by konstantin-korovin** [ocaml/ocaml#13596]
I get the following "Assembler error" with -O3 flag on M1 macOS Sonoma 14.7.1 when compiling.

(This issue does not show up on Ubuntu or some previous versions of macOS)

```
opam switch
5.2.0+flambda   ocaml-option-flambda.1,ocaml-variants.5.2.0+options   ocaml-variants = 5.2.0+options & ocaml-option-flambda

-------------------
....
ocamlfind ocamlopt -w +a-4-7-9-27-29-30-32..42-44-45-48-50-60-66..70@5@10@6+u+n+i+m+p -O3 -noassert  -I obj/ -I util/lib -package str -package yojson -package ocamlgraph -package zarith -I util/objsize-stub -package z3 -thread -o obj/theory_db.cmx -c src/theory_db.ml

/var/folders/yx/h2g1wfxj18v4jncqz_59l0dw0000gr/T/camlasmccac4f-c663b1.s:125031:2: error: fixup value out of range
 b.cs L36492

.....

File "src/theory_db.ml", line 1:
Error: Assembler error, input left in file /var/folders/yx/h2g1wfxj18v4jncqz_59l0dw0000gr/T/camlasmccac4f.s
make: *** [obj/theory_db.cmx] Error 2
-----------------

If I use -O2 in place of -O3 then the code compiles fine.

ocamlfind ocamlopt -w +a-4-7-9-27-29-30-32..42-44-45-48-50-60-66..70@5@10@6+u+n+i+m+p -O2 -noassert  -I obj/ -I util/lib -package str -package yojson -package ocamlgraph -package zarith -I util/objsize-stub -package z3 -thread -o obj/theory_db.cmx -c src/theory_db.ml

===============

The project where this occurs is 
https://gitlab.com/korovin/iprover

the module where the issue shows up is src/theory_db.ml.

To reproduce you can run:

./configure
make
===========
```

Thanks,
Konstantin

**Comment by lthls** [comment]
This looks like it would be fixed by #11162.

**Comment by konstantin-korovin** [comment]
Thanks @lthls! Indeed these issues look related. 
It looks like the (#11162) is from 2022, do you know when it maybe fixed. 

PS Also on other platforms (Linux) this module with -O3 takes considerably longer to compile compared to other modules of similar size. Might be there is an efficiency issue there as well.

**Comment by lthls** [comment]
> PS Also on other platforms (Linux) this module with -O3 takes considerably longer to compile compared to other modules of similar size. Might be there is an efficiency issue there as well.

The `Iprover` module does a lot of computations during initialisation of the module, mostly in the definition of `builtin`. Flambda tries comparatively hard to evaluate toplevel expressions at compile time, and with `-O3` you likely end up with all the calls to `add` inlined.
Since in this particular case the database initial computation is probably not performance critical, you may want to use annotations to prevent the compiler from doing that much useless work.
I would suggest either annotating `add` with `[@inline never]`, or making `builtin` a function (`let create_builtin () = ...`) and make sure the application is not inlined (`let builtin = (create_builtin[@inlined never]) ()`).

**Comment by konstantin-korovin** [comment]
Thank you!
Any tips for code rearranging that it would compile on macOS ? thanks!

(on Ubuntu) Still looks a bit strange (after the [@inline never] change) that the size of  obj/theory_db.o is 5.8M and  
src/theory_db.ml is ~500 lines (24K with comments).

**Comment by tmcgilchrist** [comment]
Just an update for @konstantin-korovin I'm working on a fix using the contents of #11162 plus a few other improvements. 
I'll have some time to finish it off next week.

**Comment by konstantin-korovin** [comment]
Excellent, many thanks @tmcgilchrist !

**Comment by tmcgilchrist** [comment]
@konstantin-korovin Do you have a branch of `iprover`  where I can verify the fix? 
Building on macOS / ARM64 I immediately run into issues using the default branch.
```
$ make
ocamlyacc src/parser_tptp.mly
6 rules never reduced
1 shift/reduce conflict.
cd src; rm -f SMTSolver.ml; ln -s SMTSolver_Z3.ml SMTSolver.ml; cd ..
./git_info_gen
git: iProver is in a git repo; saving iProver git sha1
git: sha1: 79df7f8ab196b09978d2eda20372619e4aff2caa
git: date: 2024-12-17 13:03:37 +0800
git: non-committed changes: false
if [ ! -d "./VClausifier" ]; then mkdir "./VClausifier"; fi
ln -sf ../util/vclausify_rel_mac "./VClausifier"/vclausify_rel
rm -f obj/git_info.*
./git_info_gen
git: iProver is in a git repo; saving iProver git sha1
git: sha1: 79df7f8ab196b09978d2eda20372619e4aff2caa
git: date: 2024-12-17 13:03:37 +0800
git: non-committed changes: false
cd util && /Applications/Xcode.app/Contents/Developer/usr/bin/make -f Makefile minisat-ocaml
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f /Users/tsmc/projects/iprover/util/Makefile.minisat minisat-lib-release
cd /Users/tsmc/projects/iprover/util/minisat/simp && /Applications/Xcode.app/Contents/Developer/usr/bin/make libr
Making dependencies
Depends on: core
Compiling: simp/SimpSolver.or
Compiling: utils/Options.or
Compiling: utils/System.or
Compiling: core/Solver.or
Making library: libminisat_release.a ( simp/SimpSolver.or utils/Options.or utils/System.or core/Solver.or )
a - /Users/tsmc/projects/iprover/util/minisat/simp/SimpSolver.or
a - /Users/tsmc/projects/iprover/util/minisat/utils/Options.or
a - /Users/tsmc/projects/iprover/util/minisat/utils/System.or
a - /Users/tsmc/projects/iprover/util/minisat/core/Solver.or
Making Soft Link: libminisat_release.a -> libminisat.a
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f Makefile.clib minisat-ocaml
Makefile.clib:71: warning: overriding commands for target `lib/minisat.a'
Makefile.clib:67: warning: ignoring old commands for target `lib/minisat.a'
g++ -c -O3 -o lib/minisat_stubs.o -I/Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml -I/Users/tsmc/projects/iprover/util/minisat src/minisat_stubs.cpp
In file included from src/minisat_stubs.cpp:27:
In file included from /Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/mlvalues.h:71:
In file included from /Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/domain_state.h:36:
/Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/domain_state.tbl:17:1: error: expected parameter declarator
   17 | DOMAIN_STATE(atomic_uintnat, young_limit)
      | ^
/Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/domain_state.h:35:44: note: expanded from macro 'DOMAIN_STATE'
   35 | #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
      |                                            ^
In file included from src/minisat_stubs.cpp:27:
....
```

An opam switch with the fix can be created using:
```
$ opam compiler create https://github.com/tmcgilchrist/ocaml/pull/20 --configure-command "./configure --enable-codegen-invariants"
```

**Comment by konstantin-korovin** [comment]
@tmcgilchrist Thanks, 

 ``` 
1. I created iProver branch 2025_01_17_ocaml_test (0b623c97)
where original compilation issue should show up when using 5.2.0+flambda (MacOS Sonoma v14.7.1 and Sequoia v15.2)
and I believe also this was the case for 5.2.1+flambda on Sonoma but does not show up with 5.2.1+flambda on Sequoia v15.2.  (After the MacOS upgrade I only now have Sequoia v15.2 on M1).

--------
The issue can be switch on/off if you comment/uncomment the repeated line L660 in src/theory_db.ml:
(when the line is repeated might be the alignment is restored.)

L657: let db = add_relations db in

L659: (* repeated due to a compilation issue on macOS *)  
L660: (*  let db = add_relations db in *)
 

---------

2. If you change compilers/branches you may need to run:

make clean_all
./configure
make
---------


3. It seems that the issue you see now is another error/compatibility issue in OCaml as lib/ocaml/caml/mlvalues.h and /lib/ocaml/caml/domain_state.h are OCaml C interface libraries. 

--------
In file included from /Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/mlvalues.h:71:
In file included from /Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/domain_state.h:36:
/Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/domain_state.tbl:17:1: error: expected parameter declarator
   17 | DOMAIN_STATE(atomic_uintnat, young_limit)
      | ^
/Users/tsmc/.opam/tmcgilchrist-ocaml-20/lib/ocaml/caml/domain_state.h:35:44: note: expanded from macro 'DOMAIN_STATE'
   35 | #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
      |                                            ^
In file included from src/minisat_stubs.cpp:27:
------

util/src/minisat_stubs.cpp: 27 is: 
#include <caml/mlvalues.h>

I do not see this error neither on Ubuntu nor on MacOS Sonoma/Sequoia for OCaml up to 5.2.1+flambda. 


4. When I try
   
   opam compiler create https://github.com/tmcgilchrist/ocaml/pull/20 --configure-command "./configure --enable-codegen-invariants"

I get:

[ERROR] Could not synchronize /Users/korovin/.opam/https-ocaml---github.com-tmcgilchrist-ocaml-pull-20/.opam-switch/sources/ocaml-variants from "git+https://github.com/https/ocaml#//github.com/tmcgilchrist/ocaml/pull/20":
        "/usr/bin/git fetch -q" exited with code 128
[ocaml-variants.5.4.0+trunk] fetching sources failed: git+https://github.com/https/ocaml#//github.com/tmcgilchrist/ocaml/pull/20
[ERROR] Error getting source from git+https://github.com/https/ocaml#//github.com/tmcgilchrist/ocaml/pull/20:
          - git+https://github.com/https/ocaml#//github.com/tmcgilchrist/ocaml/pull/20

Switch https-ocaml---github.com-tmcgilchrist-ocaml-pull-20 and all its packages will be wiped. Are you sure? [y/n] y
opam-compiler: internal error, uncaught exception:
               Failure("Cannot create switch - command failed: opam pin add --switch https-ocaml---github.com-tmcgilchrist-ocaml-pull-20 --yes ocaml-variants git+https://github.com/https/ocaml#//github.com/tmcgilchrist/ocaml/pull/20 --edit")
   
```

**Comment by tmcgilchrist** [comment]
That command should be as follows to get invariant checks and flambda:
```shell
opam compiler create https://github.com/ocaml/ocaml/pull/13667 \ 
--configure-command "./configure --enable-codegen-invariants --enable-flambda"
```

**Comment by konstantin-korovin** [comment]
I'm getting the following errors: 

```
1. opam compiler create https://github.com/ocaml/ocaml/pull/13667 --configure-command "./configure --enable-codegen-invariants --enable-flambda"                                                               

# To update the current shell environment, run: eval $(opam env --switch=https-ocaml---github.com-ocaml-ocaml-pull-13667)
Processing: [ocaml-variants.5.4.0+trunk: git]Username for 'https://github.com': 
Password for 'https://github.com': 
[ERROR] Could not synchronize /shareddata/home/korovin/.opam/https-ocaml---github.com-ocaml-ocaml-pull-13667/.opam-switch/sources/ocaml-variants from
        "git+https://github.com/https/ocaml#//github.com/ocaml/ocaml/pull/13667":
        "/usr/bin/git fetch -q" exited with code 128
[ocaml-variants.5.4.0+trunk] fetching sources failed: git+https://github.com/https/ocaml#//github.com/ocaml/ocaml/pull/13667
[ERROR] Error getting source from git+https://github.com/https/ocaml#//github.com/ocaml/ocaml/pull/13667:
          - git+https://github.com/https/ocaml#//github.com/ocaml/ocaml/pull/13667

Switch https-ocaml---github.com-ocaml-ocaml-pull-13667 and all its packages will be wiped. Are you sure? [y/n] y
opam-compiler: internal error, uncaught exception:
               Failure("Cannot create switch - command failed: opam pin add --switch https-ocaml---github.com-ocaml-ocaml-pull-13667 --yes ocaml-variants git+https://github.com/https/ocaml#//github.com/ocaml/ocaml/pull/13667 --edit")
 

------- it seems there are some issues with how opam-compiler create treats the url     
      
2.  opam compiler create '#13667' --configure-command "./configure --enable-codegen-invariants --enable-flambda"
Processing: [ocaml-variants.5.4.0+trunk: git]
[ERROR] Could not synchronize /Users/korovin/.opam/ocaml-ocaml-13667/.opam-switch/sources/ocaml-variants from "git+https://github.com/tmcgilchrist/ocaml#11162-rebased":
        Branch 11162-rebased not found
[ocaml-variants.5.4.0+trunk] fetching sources failed: git+https://github.com/tmcgilchrist/ocaml#11162-rebased
[ERROR] Error getting source from git+https://github.com/tmcgilchrist/ocaml#11162-rebased:
          - git+https://github.com/tmcgilchrist/ocaml#11162-rebased

Switch ocaml-ocaml-13667 and all its packages will be wiped. Are you sure? [y/n] y
opam-compiler: internal error, uncaught exception:
               Failure("Cannot create switch - command failed: opam pin add --switch ocaml-ocaml-13667 --yes ocaml-variants git+https://github.com/tmcgilchrist/ocaml#11162-rebased --edit")

```               
---- it moves further,  but it seems that this branch is gone now

@tmcgilchrist could you please send the correct branch, thanks.

**Comment by tmcgilchrist** [comment]
Please try again @konstantin-korovin the PR was merged and I deleted the branch, causing the opam failure. I've restored the branch so it should work now.

**Comment by tmcgilchrist** [comment]
```shell
$ opam compiler create ocaml/ocaml#13667 --configure-command "./configure --enable-codegen-invariants --enable-flambda"
# Run eval $(opam env --switch=ocaml-ocaml-13667) to update the current shell environment
[ocaml-variants.5.4.0+trunk] synchronised (git+https://github.com/tmcgilchrist/ocaml#11162-rebased)
[WARNING] The opam file didn't pass validation:
  warning 37: Missing field 'dev-repo'
Proceed anyway ('no' will re-edit)? [y/n] y
You can edit this file again with "opam pin edit ocaml-variants", export it with "opam show ocaml-variants --raw"
ocaml-variants is now pinned to git+https://github.com/tmcgilchrist/ocaml#11162-rebased (version 5.4.0+trunk)

The following actions will be performed:
=== install 11 packages
  ∗ base-bigarray     base
  ∗ base-domains      base
  ∗ base-effects      base
  ∗ base-nnp          base
  ∗ base-threads      base
  ∗ base-unix         base
  ∗ host-arch-arm64   1
  ∗ host-system-other 1
  ∗ ocaml             5.4.0
  ∗ ocaml-config      3
  ∗ ocaml-variants    5.4.0+trunk (pinned)

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
⬇ retrieved ocaml-config.3  (cached)
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed host-arch-arm64.1
∗ installed host-system-other.1
∗ installed ocaml-variants.5.4.0+trunk
∗ installed ocaml-config.3
∗ installed ocaml.5.4.0
∗ installed base-domains.base
∗ installed base-effects.base
∗ installed base-nnp.base
Done.
# Run eval $(opam env --switch=ocaml-ocaml-13667) to update the current shell environment
The switch invariant was set to ocaml-variants

```

**Comment by konstantin-korovin** [comment]
@tmcgilchrist I tried to complie but I'm getting OCaml errors inside OCaml C interface libraries. 
I would assume that either C interface libraries were not fully ported in this version of OCaml or the C interface
was changed in an incompatible way. (The same files compile fine on 5.2.1+flambda and below on Ubuntu).

Note that errors in the log below start from

src/minisat_stubs.cpp:27

which is (util/src/minisat_stubs.cpp)

L27: #include <caml/mlvalues.h>

and mlvalues.h is an OCaml C interface library. 

```
---------------
To reproduce you can run (from iProver branch 2025_01_17_ocaml_test)

./configure
(install libraries prompted)
make clean_all 
make
--------------
>make
ocamlyacc src/parser_tptp.mly
6 rules never reduced
1 shift/reduce conflict.
cd src; rm -f SMTSolver.ml; ln -s SMTSolver_Z3.ml SMTSolver.ml; cd ..
if [ ! -d "./VClausifier" ]; then mkdir "./VClausifier"; fi
ln -sf ../util/vclausify_rel_mac "./VClausifier"/vclausify_rel
rm -f obj/git_info.*
./git_info_gen
git: iProver is in a git repo; saving iProver git sha1
git: sha1: 0b623c972890e14f0da2a430adf76142ab742c97
git: date: 2025-01-17 15:51:05 +0800
git: non-committed changes: false
cd util && /Library/Developer/CommandLineTools/usr/bin/make -f Makefile minisat-ocaml
/Library/Developer/CommandLineTools/usr/bin/make -f /Users/korovin/iprover/util/Makefile.minisat minisat-lib-release
cd /Users/korovin/iprover/util/minisat/simp && /Library/Developer/CommandLineTools/usr/bin/make libr
Making dependencies
Depends on: core
Compiling: simp/SimpSolver.or
Compiling: utils/Options.or
Compiling: utils/System.or
Compiling: core/Solver.or
Making library: libminisat_release.a ( simp/SimpSolver.or utils/Options.or utils/System.or core/Solver.or )
a - /Users/korovin/iprover/util/minisat/simp/SimpSolver.or
a - /Users/korovin/iprover/util/minisat/utils/Options.or
a - /Users/korovin/iprover/util/minisat/utils/System.or
a - /Users/korovin/iprover/util/minisat/core/Solver.or
Making Soft Link: libminisat_release.a -> libminisat.a
/Library/Developer/CommandLineTools/usr/bin/make -f Makefile.clib minisat-ocaml
Makefile.clib:71: warning: overriding commands for target `lib/minisat.a'
Makefile.clib:67: warning: ignoring old commands for target `lib/minisat.a'
g++ -c -O3 -o lib/minisat_stubs.o -I/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml -I/Users/korovin/iprover/util/minisat src/minisat_stubs.cpp
In file included from src/minisat_stubs.cpp:27:
In file included from /Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/mlvalues.h:71:
In file included from /Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:36:
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.tbl:17:1: error: expected parameter declarator
   17 | DOMAIN_STATE(atomic_uintnat, young_limit)
      | ^
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:35:44: note: expanded from macro 'DOMAIN_STATE'
   35 | #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
      |                                            ^
In file included from src/minisat_stubs.cpp:27:
In file included from /Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/mlvalues.h:71:
In file included from /Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:36:
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.tbl:17:1: error: expected ')'
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:35:44: note: expanded from macro 'DOMAIN_STATE'
   35 | #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
      |                                            ^
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.tbl:17:1: note: to match this '('
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:35:34: note: expanded from macro 'DOMAIN_STATE'
   35 | #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
      |                                  ^
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/misc.h:152:29: note: expanded from macro 'CAMLalign'
  152 | #define CAMLalign(n) alignas(n)
      |                             ^
In file included from src/minisat_stubs.cpp:27:
In file included from /Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/mlvalues.h:71:
In file included from /Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:36:
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.tbl:17:1: error: a type specifier is required for all declarations
   17 | DOMAIN_STATE(atomic_uintnat, young_limit)
      | ^
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/domain_state.h:35:34: note: expanded from macro 'DOMAIN_STATE'
   35 | #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
      |                                  ^
/Users/korovin/.opam/ocaml-ocaml-13667/lib/ocaml/caml/misc.h:152:22: note: expanded from macro 'CAMLalign'
  152 | #define CAMLalign(n) alignas(n)
      |                      ^

```

**Comment by kit-ty-kate** [comment]
That sounds like a duplicate of https://github.com/ocaml/ocaml/issues/13422 which means that your issue should be easily fixed by adding `-std=gnu++11` to your calls to `g++`

**Comment by konstantin-korovin** [comment]
@kit-ty-kate thanks! this fixed this issue. 

@tmcgilchrist iProver compiles now on macOS Sequoia 15.3.1 so I do not get the original "Assembler error"
(I can not test on Sonoma 14.7.1, after the update, but as far as I can see all seems to work now). 

Many thanks to all!

**Comment by tmcgilchrist** [comment]
@konstantin-korovin Excellent, if you are happy with that resolution please close this issue.
Both fixes should make their way into OCaml 5.4 later this year.
